DEV: AFIDs class & file handling

kaitj opened the pull request:
## Proposed changes

Added a new AFIDs class for internal representation of fiducials rather than a JSON object. With this comes new methods and a bunch of code changes. 

Importantly, the `csv_to_json` method is now deprecated and has been removed! Thought about keeping this in, but the function is no longer used with the new class handling fiducials. Also means new functions to handle reading input files `csv_to_afids` and `json_to_afids`. Both are written similarly and followed closely to `csv_to_json`. The previous `parse_fcsv_fields` is still utilized, and a similar `parse_json_keys` has been created!

With regards to unit testing, I've updated all the previous test cases for the JSON object to work with the afids class. We may need to double check the files themselves to make sure it still makes sense with the default internal representation of the coordinate system (see below). However, without a valid fiducial JSON file, I have not written up any test cases for these types of files. Additionally, this PR may fail the Github Actions testing given the changes, but local unit testing has indicated everything is passing.

Minor things include:
* renaming of `model_auto` to `model` and similarly `test_model_auto` to `test_model`
* renamed method`InvalidFcsvError` to `InvalidFileError` to be more general
* defaulting to "LPS" internal representation
* checking if fiducial flipping is necessary by the given coordinate system rather than file version

What is not done / what I've broken include with this PR and still needs to be worked on:
* ensuring all necessary keys have values in the JSON
* test cases for JSON files (this will probably have to wait until we have some valid JSON files)
* rewriting `control.py` to make use of the new afids class

## Types of changes

What types of changes does your code introduce? Put an `x` in the boxes that apply

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionalitiy)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (if none of the other choices apply)

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you are unsure about any of the choices, don't hesitate to ask!

- [ ] Changes have been tested to ensure that fix is effective or that a feature works.
- [ ] Changes pass the unit tests
- [x] I have included necessary documentation or comments (as necessary)
- [ ] Any dependent changes have been merged and published

## Notes
All PRs will undergo the unit testing before being reviewed. You may be requested to explain or make additional changes before the PR is accepted.

_PR template was adopted from [appium](https://github.com/appium/appium/blob/master/.github/PULL_REQUEST_TEMPLATE.md)_

kaitj commented:
As a separate comment, I have created a separate branch off of `master` for a potential next release candidate (rc-0.2.x). We can discuss how this branch should be used during the next meeting, but didn't think it would be a good idea to introduce breaking changes the current stable version of the page.

@tkkuehn @greydongilmore @ostanley - Mostly created this for some code review + also looking for some help with fixing / cleaning up `control.py` and suggestions!

tkkuehn commented on controller.py:
```diff
@@ -13,7 +13,8 @@
 import numpy as np
 
-from model_auto import Average, csv_to_json, InvalidFcsvError
+from generate_visualizations import generate_visualizations
```
I don't think `generate_visualizations` exists on this branch.

kaitj replied:
I think you're right. Thought I had created this off of the master branch, but looks like I didn't.

kaitj replied:
This was fixed in the latest commit

tkkuehn commented on controller.py:
```diff
@@ -13,7 +13,8 @@
 
-from model_auto import Average, csv_to_json, InvalidFcsvError
+from generate_visualizations import generate_visualizations
+from model import Average, csv_to_json, InvalidFcsvError
```
A couple of these imports have been renamed.

kaitj replied:
Fixed renaming of these imports in latest commit

tkkuehn commented on controller.py:
```diff
@@ -405,9 +404,13 @@ def validator():
 
     result = '<br>'.join([result, msg])
 
+    visualization_html = generate_visualizations(template_data_j, user_data_j)
```
Some weirdness with this branch here - this shouldn't be added in this PR unless it's merged in after the new visualization stuff gets added to master.

kaitj replied:
Yeah, not sure what happened here (re previous comment).

kaitj replied:
So this is strange, and maybe it'll be more clear when I push the new commit with some of the changes, but none of the visualization code is in my the branch I am working with.

kaitj replied:
@tkkuehn When you take a look, can you take a second look to see if the visualization is in the latest commits? Looks like they're gone to me.

tkkuehn commented on model.py:
```diff
@@ -0,0 +1,235 @@
+from pkg_resources import parse_version
+
+EXPECTED_LABELS = [str(x + 1) for x in range(32)]
+EXPECTED_DESCS = [
```
I think we should just move everything in this module to the new `afids` module, but the `EXPECTED_DESCS` should be there at a minimum in my opinion.

kaitj replied:
Add into `afids.py` and add an import to `model.py`

tkkuehn commented on model.py:
```diff
@@ -0,0 +1,235 @@
+    submit = wtf.SubmitField(label='Submit')
+
+class InvalidFileError(Exception):
+    """Exception raised when a csv to be parsed is invalid.
```
Need to update the documentation.

kaitj replied:
Documentation updated for this class in 3616253 / 32363ac

tkkuehn commented on test/test_model.py:
```diff
@@ -1,102 +1,103 @@
+    def test_valid_fcsv(self):
         with open('test/resources/valid.fcsv', 'r') as fcsv:
-            fcsv_json = model_auto.csv_to_json(fcsv)
+            fcsv_json = model.csv_to_afids(fcsv)
```
Should probably also rename all the `fcsv_json` variables to `fcsv_afids`.

kaitj replied:
Updated in 3616253 / 32363ac

tkkuehn commented on controller.py:
```diff
@@ -13,7 +13,8 @@
 
```
There's a bunch of stuff in this module that still uses the old code and names for things, and needs to be updated.

kaitj replied:
The controller.py needs to be overhauled a bit, but I wanted to hold off on that in the initial commit as I think we mentioned we needed to clean up the code anyways. Maybe something that can be done at the same time!

tkkuehn commented on controller.py:
```diff
@@ -13,7 +13,8 @@
 
```
We'll also need to add some UI adjustments to allow users to add either .fcsv or .json files.

kaitj replied:
Is there anything else aside from setting the variable of `ALLOWED_EXTENSIONS`? I haven't updated that yet either to allow for JSON files, so everything has just been a local test on jupyter.

kaitj replied:
`json` added to the `ALLOWED_EXTENSIONS` list. If there are no other changes, this can be marked resolved!

tkkuehn commented on model.py:
```diff
@@ -0,0 +1,235 @@
+        
+        if key == "position" and (parsed_coord == 0 or parsed_coord == "RAS"):
+            value = in_json[label][key]
+            value = [value[0], value[1], value[2]]
```
Does this just keep `value` the same? Maybe I'm missing something, but it looks like it's supposed to reorder the coordinates somehow, but doesn't.

kaitj replied:
You are right. It's suppose to flip `value[0]` and `value[1]` since it is in RAS. Missed this!

kaitj replied:
Resolved in 3616253 / 32363ac. 

Added a negative sign in front of `value[0]` and `value[1]` such that the x and y coordinates should be flipped if in the wrong coordinate space.

tkkuehn commented on model.py:
```diff
@@ -0,0 +1,235 @@
+        else:
+            value = in_json[label][key]
+
+        if value is None:
```
Probably also worth checking that `value`'s  elements are real numbers, rather than strings or whatever other types exist in JSON.

kaitj replied:
Agreed. There needs to be some additional work put in to handle JSON's better. Was working off of a test file with 5 fiducials, so waiting on a valid JSON fiducial file.

tkkuehn commented on afids.py:
```diff
@@ -0,0 +1,46 @@
+
+        self.coordinate_system = coordinate_system
+
+    def set_coordinate_system(self, coordinate_system):
```
Alternatively, we can validate that the argument here is one of "LPS" or "RAS".

kaitj replied:
I think we should discuss the getter / setter methods. I agree most of them (especially the setters) are probably not going to be used, but were built in for completeness. 

To the second point, I agree and also think this would be a good place to potentially validate the coordinate system. Something to add onto to the TODO list.

kaitj replied:
Actually on additional thought, you're probably right. We don't need getter / setter methods for the coordinate systems. Will remove for now!

kaitj replied:
Resolved in 3616253 / 32363ac. 

I don't know if "validating" the coordinate system would tell us anything as it is always set to "LPS" in the code. Currently, any file that is in "RAS" is converted to "LPS" by default. We don't really have a way of truly validating the file is in the correct coordinate system aside from the description of the input file.

tkkuehn commented on afids.py:
```diff
@@ -0,0 +1,46 @@
+    def get_coordinate_system(self):
+        return(self.coordinate_system)
+
+    def add_fiducial(self, label, description, positions):
```
Would be good to validate the input here, i.e. make sure:

- `label` and `description` match (assuming we have some expected mapping across formats, I'm not sure if that's true).
- `positions` is a three-element sequence type, where each element is a real number.

And throw an exception otherwise.

tkkuehn replied:
Actually it also probably makes sense to call `self.set_fiducial_positions(label, positions)` here to reduce code duplication.

kaitj replied:
I don't think `label` and `description` would match exactly, but I think we do expect a mapping. Not sure how the label id works if a fiducial is deleted, but it should be something like `<label> - <description>` == `1 - AC` or something similar.

Agree with code duplication comment, and something to be updated!

kaitj replied:
Made use of the `self.set_fiducial_positions` method in 3616253 / 32363ac. 

Validation to come!

tkkuehn commented on afids.py:
```diff
@@ -0,0 +1,46 @@
+    def get_fiducial_description(self, label):
+        return(self.fiducials[str(label)]["description"])
+
+    def set_fiducial_positions(self, label, positions):
```
Again, probably makes sense to validate the form of `positions` then call `self.set_fiducial_position` three times.

kaitj replied:
I agree with this, though I think I was less concerned with the form of `positions` at the time of writing this and more considering if we wanted to change (e.g. flip) one spatial coordinate. I suppose we could just use the `set_fiducial_positions` to do the flip and set all 3 spatial coordinates as necessary.

ostanley replied:
So based on this convo I agree there should be validation before the position is set. Way harder to undo cause you'll need to check every number.

kaitj replied:
Been a while since visiting this, could someone refresh me on what the validation entails?

tkkuehn reviewed (commented):
A good start! I think the data structure of the `Afids` class makes sense and most of the loading code makes sense, but I've suggested a few structural changes (amalgamate the `model` and `afids` modules, and a few nitpicky code structure things. I've also noted a few things left to do that I'm sure have also been discussed elsewhere.

kaitj commented on afids.py:
```diff
@@ -0,0 +1,46 @@
+    """ Create a tree structure """
+    return(defaultdict(tree))
+
+class Afids:    
```
Agreed to this point, though I think we should hold off until we have decided on all the necessary information we want to have in the class so we can just add that in all at once.

kaitj replied:
Added a temporary `def validate()` that returns `None` for now with the comment that this method needs to be created. Will fill in once we decide on everything we want included.

kaitj commented on afids.py:
```diff
@@ -0,0 +1,46 @@
+""" Utilities to create / maintain AFIDs objects """
```
I think it makes sense as well. To that point, would it still make sense to check if the contents of the uploaded file is valid (would still need to ensure extensions are valid) or should we just have this class as the check?

tkkuehn replied:
I think it makes sense to do both, and try to catch malformed AFIDs as early as possible.

kaitj commented on afids.py:
```diff
@@ -0,0 +1,46 @@
+        self.fiducials[str(label)]["positions"]["z"] = positions[2]
+
+    def get_fiducial_description(self, label):
+        return(self.fiducials[str(label)]["description"])
```
Will do, I think i had them in parentheses when I was testing something and just forgot to remove them.

kaitj replied:
Fixed in 3616253 / 32363ac

kaitj commented:
@tkkuehn - could you take a look at the `valid_flip` test case

kaitj commented:
Few changes this morning:
1. Renamed all of the templates to use the `tpl-` prefix, which more closely aligns with BIDS (I think)
2. There was a bug in reading `fcsv` files, where it (a) wasn't flipping the fiducials when needed and (b) flipped all of the fiducial positions when it did flip. To fix (a), the `parsed_coord` should always be compared against a string (not sure if this holds true for `json`. To fix (b), only flip if the `key != "z"`. Don't believe this is an issue that the current version of the validator needs to be concerned about as it does not check the coordinate system and also handles the flip correctly.
3. The `json_to_afids` implementation is now live in this branch. There is a `valid.json` file that was converted from the `MNI2009cAsym` template in the test folder. Handling of `json` uploads required slightly different handling of the input, which makes use of `json.loads` rather than streaming through `io.streamIO` that `fcsv` was using. 
4. I reverted the `ALLOWED_EXTENSIONS` from set to list, as this was easier to determine how to handle the user upload. To that end, also refactored `allowed_file` such that it also returns the extension that is parsed.

Things that should probably tackled soon:
* Still need to implement a validation of the AFIDs class
* New test cases for `json` (similar to `fcsv`).

kaitj commented:
Added in some test cases with a new class for JSON files. Some things to discuss / address:

1. Rows == fiducials, and columns == keys. That said, do we still want to ensure all of the keys are there for each fiducial if we can extract the info we need by key?
2. Raised in issue (see #106). When adding test cases, I noticed that the order of the fiducials actually do matter now. In the JSON files there are no keys to separate the fiducials, so when parsing, an index of 0 = first fiducial, 1 = second fiducial, etc. Within each of these, there are keys that identify the fiducial by id and name, but this is unknown until it is traversed. Currently it assumes fiducials are in order and any mismatch throws an error due to incorrect expected label.
3. Invalid JSON files (whether it is content or an invalid value), raises `JSONDecodeError`. This works, but makes returning a meaningful message aside from `Invalid JSON file` difficult. There should be a way around this, but I haven't found it yet. 
4. Unit testing currently just asserts that `JSONDecodeError` is raised if there is issue loading the json file.

tkkuehn commented on controller.py:
```diff
@@ -221,264 +231,268 @@ def validator():
+                json.loads(upload.read().decode("utf-8"))
+            )
+    except json.JSONDecodeError as err:
+        result = (
+            f"Invalid JSON file ({timestamp}): {err.msg} at line "
+            f"{err.lineno}, column {err.colno}."
         )
```
This does the job of producing a descriptive error message, but it seems a little awkward to me that this kind of logic has to go here.

I think I would prefer that we unify the interface of `csv_to_afids` and `json_to_afids`, by passing both of them the string to be parsed (i.e. `upload.read().decode("utf-8")` and letting them handle it, returning an `InvalidFileError` as necessary, including when the file can't be parsed.

tkkuehn replied:
(This would also make it easier to unit test the error message we're dealing with here in `test_model.py`, where the rest of this logic is tested.)

kaitj replied:
That would probably be ideal. I cant recall at the moment, but I think i was running into issues with just passing the parsed string to `json_to_afids`. 

The decode error stems from `json.loads` when the file isn't a valid json, but agreed it is a little awkard.

kaitj reviewed (commented):
Visualization looks good. Only comment I would have is to include the units? Didn't see them in the 3D plot or the histogram.

ostanley commented on afids.py:
```diff
@@ -0,0 +1,122 @@
+
+EXPECTED_LABELS = [str(x + 1) for x in range(32)]
+EXPECTED_DESCS = [
+    ["AC"],
```
why are AC and PC not given full names as well?

kaitj replied:
Believe these are names are based off of AFIDs protocol and it was never expected for the user to save the description as "Anterior commisure". I think this is might be out of the scope of this PR and potentially the validator with respect to how the fiducials should be named.

ostanley commented on afids.py:
```diff
@@ -0,0 +1,122 @@
+        self.fiducials[str(label)]["positions"]["x"] = positions[0]
+        self.fiducials[str(label)]["positions"]["y"] = positions[1]
+        self.fiducials[str(label)]["positions"]["z"] = positions[2]
+
```
Could also just pop in a validate here to be lazy. may make things slower but I think it is worth it to check at minimum the fiducial you are changing.

ostanley commented on test/test_model.py:
```diff
@@ -0,0 +1,241 @@
+import unittest
```
do we need keys number tests before moving forward?

ostanley commented on model.py:
```diff
@@ -0,0 +1,248 @@
+    if not math.isfinite(parsed_value):
+        raise InvalidFileError(f"{field} in row {label} is not finite")
+
+    return parsed_value
```
if this function is just a check why does it return the float? doesn't matter just tripped me up.

kaitj replied:
I think at one point it was stored to a variable and that might be why it is returning something? Otherwise agree, if it is just a check, can remove the return (I think).

tkkuehn replied:
Yeah maybe there was a time when we stored the coordinates as floats, but it does seem fine for this to just return nothing.

kaitj replied:
Removed the return in 0035642

ostanley commented on controller.py:
```diff
@@ -221,264 +230,251 @@ def validator():
-            (template_x - user_x) ** 2
-            + (template_y - user_y) ** 2
-            + (template_z - user_z) ** 2
+    for element in range(1, template_afids.no_of_fiducials + 1):
```
I assume black made this spacing? If so go with it, if not it is challenging for me to read personally due to the extensive whitespace.

ostanley commented on controller.py:
```diff
@@ -221,264 +230,251 @@ def validator():
+            "{:.5f}".format(
+                np.linalg.norm(
+                    np.array(
+                        [
+                            float(pos)
+                            for pos in template_afids.get_fiducial_positions(
+                                element
+                            )
+                        ]
```
especially here

ostanley reviewed (commented):
I think it is all logical, mostly minor things. 

Major things:
* Incomplete json test set I think?
* I think the validation to put things into the object should be better.

kaitj commented:
Added the `validate()` method to both `csv_to_afids` and `json_to_afids`. I noticed this method primarily returns a boolean value, so in the event of `False` being returned, I've opted to raise an `InvalidFileError` noting that the AFIDs set is invalid. Pretty sure our other checks currently in place has some overlap and would output a more meaningful error message?

kaitj commented:
Fixed the `validate` issue. It's now been added to `controller.py` following successful `csv_to_afids` or `json_to_afids` conversion. The `result` (output message) is updated as a result of the boolean output from `validate`. Any meaningful error message will take priority!

kaitj reviewed (commented):
Looked it over quickly, it looks good to me. I'm assuming all the `True` to `False` in the markup files are because they were loaded in Slicer? Looks like its also passing unit tests!

tkkuehn commented:
Oh, I accidentally ran black on that json file so I tried to restore it by copying `valid.json` over and changing the one line -- I guess the original file was a little different.
